### PR TITLE
mongodb: handle bulk write exceptions

### PR DESCRIPTION
**Pull request by mfiedorowicz**

### CODE REVIEW

File: lib/db/mongodb/errors.go [hunk 1]
```diff
@@ -9,15 +9,16 @@ const (
...
-		return false
-	}
-
-	for _, err := range we.WriteErrors {
-		if err.Code == code {
-			return true
+	switch v := err.(type) {
+	case mongo.WriteException:
+	case mongo.BulkWriteException:
+		for _, err := range v.WriteErrors {
```
Comment (billinghamj) [inline lib/db/mongodb/errors.go:15]: Are you absolutely sure these will be the same on both? I recall there possibly being some additional wrapping for one of them (i.e. you might need to go down a level further or something)

Comment (billinghamj) [reply lib/db/mongodb/errors.go:15]: Looks 💯 to add support for them tho

Comment (mfiedorowicz) [reply lib/db/mongodb/errors.go:15]: Looks like `BulkWriteException` is being wrapped the same way as for `WriteException`: https://github.com/mongodb/mongo-go-driver/blob/cc85d3210f49d9daef608a2c61cdd3f490514e9a/mongo/integration/unified/error_test.go#L137

Now my concern is with the style: matching same block of logic within two cases one by one or duplicate the logic for each case (as in referred mongo source code).

File: lib/db/mongodb/errors.go [hunk 2]
```diff
@@ -9,15 +9,16 @@ const (
...
-	we, ok := err.(mongo.WriteException)
-	if !ok {
-		return false
-	}
-
-	for _, err := range we.WriteErrors {
-		if err.Code == code {
-			return true
+	switch v := err.(type) {
+	case mongo.WriteException:
```
Comment (Duffleman) [inline lib/db/mongodb/errors.go:13]: Golang doesn't fall through by default, you need to add the `fallthrough` key word :P  But I don't think that works on type switches :S 

Hopefully I am wrong tho

Comment (mfiedorowicz) [reply lib/db/mongodb/errors.go:13]: It does!

**Review by Duffleman — APPROVED** [review]

File: lib/db/mongodb/errors.go [hunk 3]
```diff
@@ -9,14 +9,18 @@ const (
...
+		for _, err := range v.WriteErrors {
```
Comment (billinghamj) [inline lib/db/mongodb/errors.go:14]: Could set `v.WriteErrors` on each case to a variable outside the switch, then you can avoid any duplication of the looping/code checking

File: lib/db/mongodb/errors.go (hunk 1 above)
Comment (billinghamj) [reply lib/db/mongodb/errors.go:20]: Nice, looks that way!

File: lib/db/mongodb/errors.go (hunk 3 above)
Comment (mfiedorowicz) [reply lib/db/mongodb/errors.go:14]: But then if the error is not for write or bulk write it would break?

Comment (mfiedorowicz) [reply lib/db/mongodb/errors.go:14]: Mea culpa, Mongo has only errors for `WriteException` and `BulkWriteException` 🧐

Comment (mfiedorowicz) [reply lib/db/mongodb/errors.go]: Addressed

**Review by Duffleman — APPROVED** [review]

**Pull request merged**